common/src/main/java/apoc/export/cypher/formatter/AbstractCypherFormatter.java

```diff
@@ -202,68 +202,85 @@ public String mergeStatementForRelationship(CypherFormat cypherFormat, Relations
-				Set<String> labels = getLabels(node);
-				tx.commit();
-				return new AbstractMap.SimpleImmutableEntry<>(labels, idProperties);
-			}
-		};
-		Map<Map.Entry<Set<String>, Set<String>>, List<Node>> groupedData = StreamSupport.stream(nodes.spliterator(), true)
-				.collect(Collectors.groupingByConcurrent(keyMapper));
-
-		AtomicInteger propertiesCount = new AtomicInteger(0);
-
-		AtomicInteger batchCount = new AtomicInteger(0);
-		groupedData.forEach((key, nodeList) -> {
-			AtomicInteger unwindCount = new AtomicInteger(0);
-			final int nodeListSize = nodeList.size();
-			final Node last = nodeList.get(nodeListSize - 1);
-			nodeCount.addAndGet(nodeListSize);
-			for (int index = 0; index < nodeList.size(); index++) {
-				Node node = nodeList.get(index);
-				writeBatchBegin(exportConfig, out, batchCount);
-				writeUnwindStart(exportConfig, out, unwindCount);
-				batchCount.incrementAndGet();
-				unwindCount.incrementAndGet();
-				Map<String, Object> props = node.getAllProperties();
-				// start element
-				out.append("{");
-
-				// id
-				Map<String, Object> idMap = CypherFormatterUtils.getNodeIdProperties(node, uniqueConstraints);
-				writeNodeIds(out, idMap);
-
-				// properties
-				out.append(", ");
-				out.append("properties:");
-
-				propertiesCount.addAndGet(props.size());
-				props.keySet().removeAll(idMap.keySet());
-				writeProperties(out, props);
-
-				// end element
-				out.append("}");
-				if (last.equals(node) || isBatchMatch(exportConfig, batchCount) || isUnwindBatchMatch(exportConfig, unwindCount)) {
-					closeUnwindNodes(nodeClause, setClause, uniqueConstraints, exportConfig, out, key, last);
-					writeBatchEnd(exportConfig, out, batchCount);
-					unwindCount.set(0);
-				} else {
+		// Batch stream results, process BATCH_COUNT nodes at a time
+		boolean shouldContinue = true;
+		AtomicInteger totalNodeCount = new AtomicInteger(0);
+		while (shouldContinue) {
+			AtomicInteger nodesInBatch = new AtomicInteger(0);
+			Function<Node, Map.Entry<Set<String>, Set<String>>> keyMapper = (node) -> {
+				try (Transaction tx = db.beginTx()) {
+					totalNodeCount.incrementAndGet();
+					nodesInBatch.incrementAndGet();
+					node = tx.getNodeByElementId(node.getElementId());
+					Set<String> idProperties = CypherFormatterUtils.getNodeIdProperties(node, uniqueConstraints).keySet();
+					Set<String> labels = getLabels(node);
+					tx.commit();
+					return new AbstractMap.SimpleImmutableEntry<>(labels, idProperties);
+				}
+			};
+			Map<Map.Entry<Set<String>, Set<String>>, List<Node>> groupedData =
+					StreamSupport.stream(
+							com.google.common.collect.Iterables.limit(
+									com.google.common.collect.Iterables.skip(nodes, totalNodeCount.get()),
+									exportConfig.getBatchSize()).spliterator(),
+									true)
+					.collect(Collectors.groupingByConcurrent(keyMapper));
+
+			if (nodesInBatch.get() < exportConfig.getBatchSize()) shouldContinue = false;
```

Comment:
I had to think quite a bit to realize that this was correct, so maybe it would be worth to add an explanatory code comment about the nodesInBatch counter. Same goes for `relsInBatch`